File: Dataset/Recursion_prepared/Python/0783-minimum-distance-between-bst-nodes_trace_5.py
class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right
class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        curr_smallest, prev = float("inf"), None
        
        def inorder(node):
            nonlocal curr_smallest, prev
            if node is None:
                return
            
            inorder(node.left)
            if prev is not None:
                curr_smallest = min(curr_smallest, node.val-prev.val)
            prev = node
            inorder(node.right)

        inorder(root)
        return curr_smallest

if __name__ == "__main__":
    root = TreeNode(30)
    root.left = TreeNode(20)
    root.right = TreeNode(40)
    root.left.left = TreeNode(10)
    root.left.right = TreeNode(25)
    root.right.left = TreeNode(35)
    root.right.right = TreeNode(50)
    solution = Solution()
    print(solution.minDiffInBST(root))

File: Dataset/Advanced/Test/Recursion_prepared/Java/0783-minimum-distance-between-bst-nodes_trace_5.java
class Solution {
    TreeNode prev = null;
    int res = Integer.MAX_VALUE;

    public int minDiffInBST(TreeNode root) {
        dfs(root);
        return res;
    }

    private void dfs(TreeNode node) {
        if (node == null) {
            return;
        }
        dfs(node.left);
        if (prev != null) {
            res = Math.min(res, node.val - prev.val);
        }
        prev = node;
        dfs(node.right);
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        TreeNode root = new TreeNode(15);
        root.left = new TreeNode(10);
        root.right = new TreeNode(20);
        root.left.left = new TreeNode(8);
        root.left.right = new TreeNode(12);
        root.right.left = new TreeNode(17);
        root.right.right = new TreeNode(25);
        System.out.println(solution.minDiffInBST(root));
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}